### PR TITLE
Up button revamp

### LINKED ISSUE #81: Up button visibility & text
Hi Team,

I was going to suggest a back button for navigation up the directory structure until I discovered you already have one. However by default it is invisible until it is hovered over, not sure if that was the intention?

viv_parent_link

```css
a#parent-dir {
    position: fixed;
    height: 3rem; width: 3rem;
    line-height: 3rem; text-align: center;
    background-color: #333;
    border-radius: 9999px;
    left: 2rem; top: 2rem;
    opacity: 0;
}
a#parent-dir:hover { opacity: 1; }
```

### PR DESCRIPTION
**Pull request by Tweekism**
Close #81 

Hey mate,

Not sure if you want it, but I've set mine up like this. I mean we can always make it more pretty later right?

image

### CODE REVIEW

**Comment by Tweekism** [comment]
Oh I thought maybe you'd prefer the `dirParent` function in the `utils/path.ts` file, so can move it if you prefer. Or just tell me to go away lol.

**Comment by jannis-baum** [comment]
Thanks for the PR! I like the idea of having the option to include the name of the parent directory. I would wrap it in a span with a class name so that custom CSS can hide it again though, just keeping options open for the user :)

Let me think about this for a bit and then I'll make a suggestion for something in between this and what we had before

**Comment by Tweekism** [comment]
I'm no HTML5 / CSS expert, could you do something with data attributes like:-

```html
<a id="parent-dir" data-parent-folder="${dirParent(path)}" href="${pathToURL(pdirname(path))}" />
```
And then the user in CSS could go something like:-

```css
a#parent-dir { content: "🠈 "attr(data-parent-folder) ; }
```

**Comment by jannis-baum** [comment]
> I'm no HTML5 / CSS expert, could you do something with data attributes like:-
> 
> ```
> <a id="parent-dir" data-parent-folder="${dirParent(path)}" href="${pathToURL(pdirname(path))}" />
> ```
> 
> And then the user in CSS could go something like:-
> 
> ```css
> a#parent-dir { content: "🠈 "attr(data-parent-folder) ; }
> ```

This makes you more of a CSS/HTML5 expert than me, lol. I didn't know about this, looks very cool!

**Comment by Tweekism** [comment]
Yeah it didn't work when i tried it tho lol.

**Comment by Tweekism** [comment]
Ok it actually does work, but I noticed there are 3 `<a>` tags with the `parent-dir` id

image

Solutions would be remove the extra 2 `<a>` so there is only one `parent-dir` id, or make `parent-dir` a class and make a new id `back-link` just for the back button

This otherwise solves the the issue of letting the default style be exactly as it was while also letting custom css do its magic.

Let me know your thoughts.

**Comment by Tweekism** [comment]
Ok so here is what I have...

For the default style sheet, nothing layout wise has changes except by default the opacity is now 30%. You can set this back to zero in your custom css

As seen below that is just barely visible on the dark theme and we can tweak this over time of course and per theme with the `@media (prefers-color-scheme: light)` selectors.

Screenshot from 2024-07-14 00-50-10Screenshot from 2024-07-14 00-50-20

With cursor hovering over, opacity is 100% as before...

Screenshot from 2024-07-14 00-50-50Screenshot from 2024-07-14 00-51-02

AND then, with just a little bit of custom css you can do this...

```css
a#parent-dir {
    position: static;
    background-color: transparent;
    opacity: 1;
}

a#parent-dir::after {
    content: "🠄 "attr(data-parent-folder);
}
```
Screenshot from 2024-07-14 01-33-07Screenshot from 2024-07-14 01-32-58

Let me know what you think.

**Comment by Tweekism** [comment]
Also let me know if you need me to change any of those commit messages or move that helper function to utils etc.

**Comment by jannis-baum** [comment]
Nice, thanks! Looks cool on first sight! I'll try and check it out tomorrow. Until then it would be great if you could rebase onto main (we just merged another PR) and edit your second commit message to be "conventional" as well, e.g. `feat(#82): option for name of parent directory in back button` :)

**Comment by Tweekism** [comment]
Hey mate,

I've made those changes but i haven't been able to test it properly yet. When all your changes came down something happened to my `yarn dev` thing and it broke. Going to go have a look at that now.

**Comment by Tweekism** [comment]
Yeah UI design is clearly not my thing.

I do think we'll need to wrap the back link in a div or something, to contain it and enable more layout options. Then you can setup your max-width to match the body-content etc.

something like:-

```css
div#parent-dir-content {
    max-width: 900px;
    margin: auto;
    text-align: right;
}

a#parent-dir {
    position: static;
    background-color: transparent;
    display: static;
    opacity: 1;
}

a#parent-dir::after {
    content: "🠄 "attr(data-parent-folder);
}
```
Screenshot from 2024-07-14 16-01-27

**Comment by jannis-baum** [comment]
> Hey mate,
> 
> I've made those changes but i haven't been able to test it properly yet. When all your changes came down something happened to my `yarn dev` thing and it broke. Going to go have a look at that now.

Oh, sorry about that! Were you able to resolve this? Maybe it's just running just `yarn` one time to update your locally installed packages according to the changes?
 
> Edit: Also wasn't sure what your naming convention is for functions and what not. Most look like a simple `verbNoun` pair so I went with that.

We don't have anything super fixed, what you say sounds good! Do you think this is something we should define, e.g. in the contribution guide?

**Comment by jannis-baum** [comment]
By the way I think this `attr(data-parent-folder)` thing is not something people will figure out by themselves ^^ So I think we need some way of documenting/suggesting it.

First thought that comes to my mind would be opening a pinned "customization" discussion or something where people can share stuff they did to/with their Vivify. What do you think?

**Comment by Tweekism** [comment]
> Oh, sorry about that! Were you able to resolve this? Maybe it's just running just `yarn` one time to update your locally installed packages according to the changes?

Yeah I figured it out pretty quick, I needed to bring down your new dependencies and something just got stuck. A reboot sorted it out though.

> We don't have anything super fixed, what you say sounds good! Do you think this is something we should define, e.g. in the contribution guide?

No I don't think so at this stage. No one wants to read 20 pages of rules before they can look at the code. Once the code grows at bit there will be more examples to look at. 

> By the way I think this `attr(data-parent-folder)` thing is not something people will figure out by themselves ^^ So I think we need some way of documenting/suggesting it.
> 
> First thought that comes to my mind would be opening a pinned "customization" discussion or something where people can share stuff they did to/with their Vivify. And then linking to it from the README. What do you think?

So I've been thinking about this, I'm actually leaning towards your original idea and just putting the back icon and text into separate spans and calling it a day. 

Then users can just toggle them on or off in CSS making the 90% of use cases a lot more discoverable. Those in the know can still use ::before and ::after to insert a custom icon if they want to, and maybe we document that.

I'd like to avoid the trap of over complicating things with extra customization options that no one really uses as each one comes with an additional maintenance overhead (I'm looking at you KDE devs).

Maybe we could include an example or two custom css files somewhere for people to look at?

**Comment by Tweekism** [comment]
Oh also, could you let me know which back button style you were wanting to use for the default.

I assumed it was the current fixed position button in the corner. I have an uncommitted change here that cleans up the look a little bit for it (see below).

Or I could set the default to my text based version that is aligned with the body text at the top of the page.

Could you let me know which one you want?

Screenshot from 2024-07-14 19-58-27Screenshot from 2024-07-14 19-58-19

Or the text based (you've seen before) there is still some tightening up of the layout to be done here but you get the general idea)

Screenshot from 2024-07-14 20-07-25

**Comment by jannis-baum** [comment]
> Could you let me know which one you want?

I think the more sensible default is to have the text displayed. But I think I would prefer having it all the way on the left like the button we had before instead of aligned with the document.

Since it will then always be visible let's also hide the button if we're already at the root directory.

By the way, I like this arrow that you have in your example but what kind of character is it? If it's this one `🠄 ` (copied from your code sample) I think we have to use a more standard one because it doesn't display on my system: GitHub shows empty space, in other places I get a question mark in a box.

**Comment by Tweekism** [comment]
> I think the more sensible default is to have the text displayed. But I think I would prefer having it all the way on the left like the button we had before instead of aligned with the document.

Ok, I will have to lock in some space for it and will need to tidy up how the wrapping looks, otherwise it overlaps on to the body text when there are longer directory names on smaller monitors.
 
> ...let's also hide the button if we're already at the root directory.
 
Can do, I was limiting my changes to the CSS, but I'm happy to do it.

> By the way, I like this arrow that you have in your example but what kind of character is it? If it's this one `🠄 ` (copied from your code sample) I think we have to use a more standard one because it doesn't display on my system: GitHub shows empty space, in other places I get a question mark in a box.

I just picked one at random that looked nice, its [this one](https://www.compart.com/en/unicode/U+1F804) from the unicode specification but its way down the list. Specifically the Unicode 7.0 spec from 2014. I'll look for a more compatible one.

I assume there is no rush for this on your end? I'd like to just poke at it now for a couple days after work and stuff...

**Comment by jannis-baum** [comment]
> Ok, I will have to lock in some space for it and will need to tidy up how the wrapping looks, otherwise it overlaps on to the body text when there are longer directory names on smaller monitors.

True, I didn't consider this. Then I guess let's just do what you had in the screenshot ^^ Might look weird otherwise.

> I just picked one at random that looked nice, its [this one](https://www.compart.com/en/unicode/U+1F804) from the unicode specification but its way down the list. Specifically the Unicode 7.0 spec from 2014. I'll look for a more compatible one.

Yeah, weird, I don't know why Apple doesn't like that one haha. If I [search for "leftwards arrow"](https://www.compart.com/en/unicode/search?q=leftwards+arrow#characters) everything from `U+1F800` and after as well as `U+2BA8` and `U+2BAA` aren't compatible for me. All the others work :)

> I assume there is no rush for this on your end? I'd like to just poke at it now for a couple days after work and stuff...

No rush at all, take your time! We will hopefully release [version 0.2.0](https://github.com/jannis-baum/vivify/milestone/2) soon but if this isn't finished by then we can just do 0.2.1 :)

**Comment by tuurep** [comment]
I see the `🠄` arrow, by the way :D

Maybe that depends on your font?

**Comment by Tweekism** [comment]
Yeah it'll be a font thing, unicode has got so big and complicated with so many code points font makers can't make them all.

I'll probably make a little .svg to put there. 😅

**Comment by tuurep** [comment]
On Linux I've usually fixed problems like this by installing fonts that cover a wide range of unicode/emoji that act as fallback fonts

Specifically Noto, Noto CJK and Noto Emoji

**Comment by jannis-baum** [comment]
> Maybe that depends on your font?
>
> On Linux I've usually fixed problems like this by installing fonts that cover a wide range of unicode/emoji that act as fallback fonts

Yes, probably. I think on macOS I can't pick a system font lol. But it's for the better; if I used Linux I would never leave the house again because the only thing I would do is customize my computer hahahaha it's already bad enough even with the restrictions I have on macOS

> I'll probably make a little .svg to put there. 😅

I like the dedication! haha

**Comment by tuurep** [comment]
Can confirm not leaving the house for that reason but also shocked that you can't change the font :smile:

**Comment by jannis-baum** [comment]
A bit off-topic but @Tweekism's thought on the SVG

> I'll probably make a little .svg to put there. 😅

made me think about the emojis we are currently using as file/directory icons. These would also look so much nicer if they were more minimal SVGs, maybe like what GitHub uses for their directory lists. @Tweekism is this something that would also interest you?🙈

**Comment by Tweekism** [comment]
I'm sure there are some existing icon sets out there with an appropriate license we could use. 

Did you want me to have a look?

**Comment by jannis-baum** [comment]
> I'm sure there are some existing icon sets out there with an appropriate license we could use.
> 
> Did you want me to have a look?

That would be great! But only if it interests you and you would like doing it :) If not I can look for something

**Comment by Tweekism** [comment]
Yeah, I'll have a look when I get back, I'm heading away for work for the next 2 days. I was hoping to finish the updates to the contributing docs before I left but I've run out of time now.

**Comment by jannis-baum** [comment]
Hey I took a quick look and saw that [GitHub's icons](https://github.com/primer/octicons?tab=readme-ov-file) are MIT-licensed and seem to be very accessible. So that might be an option for the icon :) Another could be [Feather](https://feathericons.com), also MIT-licensed.

**Comment by Tweekism** [comment]
OK, so finally getting back to this and I have some thoughts to run past you guys.

So @jannis-baum I noticed you moved the "max-width" from the body section to the "content-*type*" sections. Meaning the back button doesn't align with the content anymore.



Unless the button is moved inside the content sections. 

In which case you could customise the button per content type. For example the back button on the directory parser might be:- 

image

And then the Markdown back button could be:-

image

BUT Then.... 

If we are going to do all that, maybe the current global back should stick around in some form? perhaps as an option togglable in the config?

Thoughts me up boys!

PS. Technically this is a "Up" button rather than a "Back" button as it navigates up the directory structure and not necessarily back where you came from. I'm calling it a back button here but in the code maybe we should use the correct term?

**Comment by jannis-baum** [comment]
> So @jannis-baum I noticed you moved the "max-width" from the body section to the "content-type" sections. Meaning the back button doesn't align with the content anymore.

True! I didn't consider that change would affect this PR as well.

> Unless the button is moved inside the content sections.
> 
> In which case you could customise the button per content type. For example the back button on the directory parser might be:-
> 
> And then the Markdown back button could be:-

Oh, I really like these ideas! Both are great! Only things we will have to consider are

1. Where is the button when there isn't enough space next to the heading/whatever is in the top of the Markdown file?
2. What about the other text rendering types (Notebooks & plain text/code)?

> If we are going to do all that, maybe the current global back should stick around in some form? perhaps as an option togglable in the config?

Not sure I understand this correctly. Do you mean having two buttons that are there at the same time? If so, I would prefer not doing this and just having the new ones.

> PS. Technically this is a "Up" button rather than a "Back" button as it navigates up the directory structure and not necessarily back where you came from. I'm calling it a back button here but in the code maybe we should use the correct term?

Good point! Yes, definitely should start using the correct term from now on and make code adjustments in case we call it "back button" anywhere at the moment.

**Comment by jannis-baum** [comment]
I don't know how much you have looked into the code so far but place to add the up button is probably in `src/parser/parser.ts` and in there maybe in the `wrap` function, e.g. based on an optional `parentDir` argument so that you can provide it for the files and omit it for the directories and add the directory-specific style directly in `renderDirectory` instead.

**Comment by Tweekism** [comment]
> Not sure I understand this correctly. Do you mean having two buttons that are there at the same time? If so, I would prefer not doing this and just having the new ones.

I think @tuurep likes the current button.

So everything I've done so far has been with the mindset that it would be trivial, with a couple of lines of css, to set the "up button" back to the style we have now. 

If we go down this route, that would no longer be the case.

However, if you are really attached to the old button, it would still be pretty easy to make your own branch and add it in the HTML.

If you guys are happy with these assumptions I'll start working on it in earnest.

**Comment by jannis-baum** [comment]
> I think @tuurep likes the current button.
> 
> So everything I've done so far has been with the mindset that it would be trivial, with a couple of lines of css, to set the "up button" back to the style we have now.
> 
> If we go down this route, that would no longer be the case.
> 
> However, if you are really attached to the old button, it would still be pretty easy to make your own branch and add it in the HTML.
> 
> If you guys are happy with these assumptions I'll start working on it in earnest.

Ah, okay! I don't think you have to worry about this to be honest. Just implement it as it makes sense and in a way the code is clean and don't think about the old button at all.

We have config options for including custom CSS and JS, and the custom JS has access to the path that the viewer is showing. So it's definitely possible to create any kind of custom back button now matter how you implement this one :)

**Comment by Tweekism** [comment]
> We have config options for including custom CSS and JS, and the custom JS has access to the path that the viewer is showing. So it's definitely possible to create any kind of custom back button based on just the config, no matter how you implement this one :) No need for a extra branch or anything

That makes it much easier, thanks :)

**Comment by Tweekism** [comment]
Whoopsie, I didn't know renaming the branch would close the PR, will fix after lunch.

**Comment by Tweekism** [comment]
Reopen after accidental close

**Comment by Tweekism** [comment]
Ok there is enough here now for anyone who wants to have a look.

Note that I haven't removed the old button yet, will do that before merge. 

This is the wrapping behaviour, it now wraps on contact with the link (plus a small margin) instead of the edge of the content div, otherwise it is the same.

Screencast from 2024-07-28 16-53-28

This is enabled for markdown, notebooks, and text rendering.

And this is the directory listing

image

It doesn't have a custom icon yet.

@jannis-baum we talked about swapping these out for some .svg icons. With your permission, I'd like to work on that next as its own issue? Its a huge boon to my motivation to get smaller chunks completed / merged more often.

Not being a web dev guy, everything takes way longer than it should. I probably spend equal time fixing my tooling and pacing around my office as I do actually contributing to the issue haha.

Oh and btw, can you both let me know if you can render this character in your browsers? -->   ◂

Should look like this:-
image

**Comment by Tweekism** [comment]
@jannis-baum Oh, and please let me know what javascript/typescript sins I have made :)

**Comment by jannis-baum** [comment]
Hello! This looks really good on your screenshots! I like the up button for directories as it is but I think the one for files only works well when the first thing on the file is a headline.

If I start a Markdown file with a paragraph it's a bit off

image

and for a Notebook starting in a code cell as well

image

Maybe we could put it above the content as you had it at some point? What do you think?

---

> @jannis-baum we talked about swapping these out for some .svg icons. With your permission, I'd like to work on that next as its own issue? Its a huge boon to my motivation to get smaller chunks completed / merged more often.

Yes, definitely! 100% with you on this. Feel free to work on #102 next if you want :)

> Not being a web dev guy, everything takes way longer than it should. I probably spend equal time fixing my tooling and pacing around my office as I do actually contributing to the issue haha.

Haha yes, I get this, I usually also try to only touch web development with a 3m pole while closing my eyes hahaha, for Vivify it just seemed like the sensible thing to do😂

> Oh and btw, can you both let me know if you can render this character in your browsers? -→ ◂

Works for me!

> @jannis-baum Oh, and please let me know what javascript/typescript sins I have made :)

I'll have a look at the code later😁

**Comment by Tweekism** [comment]
Yeah I noticed that on one of my notes (markdown notes) that didn't have a header. I don't have a plan how fix that yet, will need to think about it.

The other bug, and you should test this on your browser, is on .ipynb notebooks the link isn't clickable, seems like something is covering it? If you unfloat it it becomes clickable......

**Comment by jannis-baum** [comment]
> Yeah I noticed that on one of my notes (markdown notes) that didn't have a header. I don't have a plan how fix that yet, will need to think about it.

Oki no worries. I also liked your original idea of putting it above the start of the content. You don't like that one anymore?

> The other bug, and you should test this on your browser, is on .ipynb notebooks the link isn't clickable, seems like something is covering it? If you unfloat it it becomes clickable......

Ah, yes, I hadn't tried that. It's probably because of the `position: relative` that the code cells have. Anyways, there should be a better way than to use `float` as this is considered a legacy technique now: https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Floats#the_background_of_floats

**Comment by Tweekism** [comment]
> Oki no worries. I also liked your original idea of putting it above the start of the content. You don't like that one anymore?

Ok lets just to that, thats easy. Do you like left or right better?

image

image

putting it down in the header is difficult because thats rendered by the external library, and then as you noticed, you are not always guaranteed to even have a header.

I guess you could do some javascript to detect if the first line was a header or not and then position accordingly, but for now, i'm happy with this...

**Comment by jannis-baum** [comment]
> Ok lets just to that, thats easy. Do you like left or right better?

Sounds good! I'd say left. I think that's the more conventional location where people would expect an up/back button

**Comment by tuurep** [comment]
> Oh and btw, can you both let me know if you can render this character in your browsers? --> ◂

Yes this renders for me

SVG icons should be good and straight forward right now, if we have agreed on the Octicons and maybe color them like GitHub does. Once again, what dark theme colors...? :smile:

**Comment by jannis-baum** [comment]
> SVG icons should be good and straight forward right now, if we have agreed on the Octicons and maybe color them like GitHub does. Once again, what dark theme colors...? 😄

Let's discuss in #102!

**Comment by Tweekism** [comment]
Just playing few games with some friends at the moment, in about 30 mins I'll go de-float this 😊

**Comment by Tweekism** [comment]
Ok nav is back at top, defaults to the left.

I'm weird, kinda like it on the right, so I wrapped it in a <div> to make it easy to move around in custom .css.

Also have removed the old / original button.

**Review by jannis-baum — CHANGES_REQUESTED** [review]
Looks good. Just that small thing but afterwards we can merge :)

File: src/parser/parser.ts
Comment (Tweekism) [reply src/parser/parser.ts:16]: LMAO, I knew you'd pick that up.

I nearly made it a ternary on one one line reaching out to column 200 or something.

So what the best way to do this, there is a commit button here, or is it better if I make the change locally, test it and push up a new commit?

Comment (jannis-baum) [reply src/parser/parser.ts:16]: > LMAO, I knew you'd pick that up.
> 
> I nearly made it a ternary on one one line reaching out to column 200 or something.

Hahahaha nice I actually love to do stuff like that😂 But here I try to hold back because I want it to be accessible to others

> So what the best way to do this, there is a commit button here, or is it better if I make the change locally, test it and push up a new commit?

The latter ^^

Comment (Tweekism) [reply src/parser/parser.ts:16]: I noticed, I seriously considered it, and I used just the shortened if matching against the empty string but removed it for readability. Don't know why I then didn't think to use that method on the expanded if statement.

If you don't mind me rewriting history I can squish this into the original commit and force push it up.

File: src/parser/parser.ts
Comment (jannis-baum) [inline src/parser/parser.ts:19]: Actually, let's also omit the `◂ ` part here and instead handle it in the CSS, e.g. with something like

```css
#top-nav > a:before { content: '◂' }
```

This way people can easily customize and adjust that part

File: src/parser/parser.ts
Comment (jannis-baum) [reply src/parser/parser.ts:16]: > If you don't mind me rewriting history I can squish this into the original commit and force push it up.

No strong opinion, do as you please :)

File: src/parser/parser.ts
Comment (Tweekism) [reply src/parser/parser.ts:19]: Can do.

There is a chance this will change again when the .svg icons go in.

Comment (jannis-baum) [reply src/parser/parser.ts:19]: For reference you can look at this

https://github.com/jannis-baum/Vivify/blob/ba76b52bb46f21b053d9c3dc9eeb2da50d909ced/static/style.css#L106-L108

> There is a chance this will change again when the .svg icons go in.

Yes, probably, I'd just do it like this now anyways. Maybe the SVG icons can also be added this way

Comment (Tweekism) [reply src/parser/parser.ts:19]: Yeah fair enough.

Actually I just realised its now 1:30 am and I do have work tomorrow.

I'll better head off and I'll do this in the morning.

Comment (jannis-baum) [reply src/parser/parser.ts:19]: No rush! Sleep is definitely the priority ^^

**Review by jannis-baum — APPROVED** [review]

**Comment by jannis-baum** [comment]
Good work! Thank you!

**Comment by tuurep** [comment]
Sorry to bother after merge but gonna give an opinion on this:

image

1. `..` directory a nice idea for dir list
2. I think that `(parent)` looks redundant as the same info is always on the path on the top. Also looks quite unholy.

The benefit of it is that the `..` is a tiny target to click and that makes it bigger

But what I would do instead is make the directory icon clickable as part of the link (which isn't currently)

**Comment by jannis-baum** [comment]
> Also looks quite unholy.

Hahaha yes a bit.

> But what I would do instead is make the directory icon clickable as part of the link (which isn't currently)

This sounds good! Maybe we do it in the context of #102? Or do you think it's more urgent and should happen ASAP? If so feel free to open a PR and patch it ^^

**Comment by tuurep** [comment]
Yeah with #102 is totally fine!

**Comment by Tweekism** [comment]
> The benefit of it is that the `..` is a tiny target to click and that makes it bigger
> 
> But what I would do instead is make the directory icon clickable as part of the link (which isn't currently)

You summed it up quite perfectly. I never actually said it, but that was my plan.

**Pull request merged**
